### PR TITLE
cluster_management/LeaderFollower: use Helix assigned leader as upstream before OFFLINE->FOLLOWER transition completes

### PR DESCRIPTION
**Pull request by jaricftw**
Upon OFFLINE->FOLLOWER state transition, the db tries to find an upstream to catch up:
- if leader is unavailable, the db will find a live `follower` to catch up, and use it as the upstream. 
- if no healthy leader/follower is available, set `itself` as upstream

This is problematic, because the follower would continue to use a `follower` or `itself` as the upstream after the state transition completes. 

Fixing that by using the Helix-assigned true leader as the upstream, after the replica has tried to catch up (maybe from a follower), before the state transition ends.

### CODE REVIEW

**Comment by kangnanli** [comment]
Can you add details:
- when will a follower has another follower as its upstream?

And, your current PR will not fix this case, bcz, this PR only change the logic at transition: Offline -> Follower. It does not impact the Follower itself.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/LeaderFollowerStateModelFactory.java [hunk 1]
```diff
@@ -385,6 +385,18 @@ public void onBecomeFollowerFromLeader(Message message, NotificationContext cont
       partitionStateUpdater.removeLeader(resourceName, partitionName);
     }
 
+    public String getLeaderInstance(NotificationContext context) {
+      HelixAdmin admin = context.getManager().getClusterManagmentTool()
+      ExternalView view = admin.getResourceExternalView(cluster, resourceName);
+      Map<String, String> stateMap = view.getStateMap(partitionName);
```
[the remark this answers is not in the document]
Comment (jaricftw) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/LeaderFollowerStateModelFactory.java:391]: yeah it's a little unfortunate. I haven't seen a way to do that: https://github.com/apache/helix/blob/a3b6a4dd1d485c1c4ef16644ff4695572e253bbf/helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAccessor.java#L524. 

That said, this is just one additional check in the state transition. we have already been doing this in lots of places, including some places where we would check external view every 1s for minutes, so I don't think it's adding significant overhead for this case. 

We could reuse the ExternalView check earlier in the state transition, but I am doing this again mostly to get a fresh view to reduce the chance of setting a wrong upstream.

**Comment by jaricftw** [comment]
> Can you add details:
> 
> * when will a follower has another follower as its upstream?
> 
The summary briefly mentioned this: "if leader is unavailable". This could happen due to various errors, e.g. unreachable leader (leader down), leader is overloaded (load shedding). I have seen both cases in production. 

> And, your current PR will not fix this case, bcz, this PR only change the logic at transition: Offline -> Follower. It does not impact the Follower itself.

Not sure I understand, this PR ensures the replica sets the correct upstream before becoming an follower, so it'll call `changeDBRoleAndUpStream` to change the follower to use the correct upstream. https://github.com/pinterest/rocksplicator/blob/master/rocksdb_admin/admin_handler.cpp#L1443-L1455

**Comment by kangnanli** [comment]
> 

I got it wrong. 
I thought follower will set its upstream to another follower. 
But, the actually upstream setup is right before it becomes a Follower. Which answer both parts of my question above.

**Review by kangnanli — APPROVED** [review]

**Comment by newpoo** [comment]
I vaguely remember that when a node first becomes the leader, it will try
to set all followers replicate from itself. Is this the case?

**Comment by jaricftw** [comment]
> I vaguely remember that when a node first becomes the leader, it will try to set all followers replicate from itself. Is this the case?

Yes, I think so: https://github.com/pinterest/rocksplicator/blob/master/cluster_management/src/main/java/com/pinterest/rocksplicator/LeaderFollowerStateModelFactory.java#L318. 

However, that's a best effort, based on what's listed as followers in the shard map. In addition, it doesn't continuously set other followers after a leader is promoted and running, so newly added followers may use the wrong leader and the true leader is not aware.

**Pull request merged**